File: neetcode150/easy/1-Two-Sum.py
#!/usr/bin/python3

class Solution:
  def twoSum(self, nums: list, target: int) -> list:
    # Initialize a dictionary
    m = {}

    # enumerate build-in function keeps track of iterations
    # which we use to represent the index of each number from
    # the list.
    for i, num in enumerate(nums):
      # Check if the number is already in the dictionary
      if num in m:
        # Return the indexs at the first value 
        return [m[num], i]
      # Store the difference between the target and number to
      # keep track of the number needed to solve the problem.
      m[target - num] = i

File: neetcode150/easy/cpp/1-Two-Sum.cpp
#include <vector>
#include <map>
using namespace std;

// Solution 1
class Solution{
public:
  vector<int> twoSum(vector<int>& nums, int target){
    map<int, int> m;
    for(int i = 0; i < nums.size(); i++){
      if(m.find(nums[i]) == m.end()){
        m[target - nums[i]] = i;
      } else{
        return vector<int>({m[nums[i]], i});
      }
    }    
    return vector<int>({-2, -1});
  }
};